lib/java_buildpack/jre/ibm_jre_initializer.rb

```diff
@@ -0,0 +1,228 @@
+# Cloud Foundry Java Buildpack
+# Copyright 2017 the original author or authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'digest'
+require 'fileutils'
+require 'java_buildpack/component/versioned_dependency_component'
+require 'java_buildpack/jre'
+require 'java_buildpack/util/tokenized_version'
+
+module JavaBuildpack
+  module Jre
+
+    # Encapsulates the detect, compile, and release functionality for selecting a JRE.
+    class IbmJreInitializer < JavaBuildpack::Component::VersionedDependencyComponent
+
+      # Creates an instance
+      #
+      # @param [Hash] context a collection of utilities used the component
+      def initialize(context)
+        @application    = context[:application]
+        @component_name = context[:component_name]
+        @configuration  = context[:configuration]
+        @droplet        = context[:droplet]
+
+        @droplet.java_home.root = @droplet.sandbox + 'jre/'
+      end
+
+      # (see JavaBuildpack::Component::BaseComponent#detect)
+      def detect
+        @version, @uri             = JavaBuildpack::Repository::ConfiguredItem.find_item(@component_name,
+                                                                                         @configuration)
+        @droplet.java_home.version = @version
+        super
+      end
+
+      # (see JavaBuildpack::Component::BaseComponent#compile)
+      def compile
+        download(@version, @uri['uri'], @component_name) do |file|
+          check_sha256(file, @uri['sha256sum'])
+          with_timing "Installing #{@component_name} to #{@droplet.sandbox.relative_path_from(@droplet.root)}" do
+            install_bin(@droplet.sandbox, file)
+          end
+        end
+        @droplet.copy_resources
+      end
+
+      # (see JavaBuildpack::Component::BaseComponent#release)
+      def release
+        @droplet
+          .java_opts
+          .add_system_property('java.io.tmpdir', '$TMPDIR')
+        @droplet.java_opts << '-Xtune:virtualized'
+        @droplet.java_opts.concat mem_opts
+        @droplet.java_opts.concat tls_opts
+        @droplet.java_opts << '-Xshareclasses:none'
+      end
+
+      private
+
+      # constant HEAP_RATIO is the ratio of memory assigned to the heap
+      # as against the container total and is set using -Xmx.
+      HEAP_RATIO = 0.75
```

Comment:
These two should be private constants.  Ruby sadly, doesn't make constants defined after `private`, private.  You'll need an invocation, after the definitions of something like `private_constant :HEAP_RATIO, KILO`, even though they follow the `private` declaration.